src/connectivity.jl

```diff
@@ -1,26 +1,32 @@
 # Parts of this code were taken / derived from Graphs.jl. See LICENSE for
 # licensing details.
 """
-    connected_components!(label, g)
+    connected_components!(label, g, [search_queue])
```

Comment:
I am all for performance improvements. But I am a bit skeptical if it is worth making the interface more complicated.

Almost all graph algorithms need some kind of of work buffer, so we could have something like in al algorithms but in the end it should be the job for Julia's allocator to verify if there is some suitable piece of memory lying around. We can help it by using `sizehint!` with a suitable heuristic.

Reply:
I agree that this will usually not be relevant; in my case it is though, and is the main reason I made the changes. I also agree that there is a trade off between performance improvements and complications of the API. On the other hand, I think passing such work buffers as _optional_ arguments is a good solution to such trade-offs: for most users, the complication can be safely ignored and shouldn't complicate their lives much.

As you say, there are potentially many algorithms in Graphs.jl that _could_ take a work buffer; in light of that, maybe this could be more palatable if we settle on a unified name for these kinds of optional buffers, so that it lowers the complications by standardizing across methods.
Maybe just `work_buffer` (and, if there are multiple, `work_buffer1`, `work_buffer2`, etc?)

Reply:
If we do this then all functions should take exactly one `work_buffer` (possibly a tuple) and have an appropriate function to initialize the buffer. I think it is a major change which should be discussed separately.
What is your use case @thchr ?

Reply:
- https://github.com/JuliaGraphs/Graphs.jl/issues/410

Reply:
So I think if this is really important for your use case you can either
- Create a version that uses a buffer in the `Experimental` submodule. Currently we don't guarantee semantic versioning there - this allows use to remove things in the future without breaking the API.
- Or as this code is very simple you might just copy it to your own repository.


But just to clarify - your problem is not that you are building graphs by adding edges until they are connected? Because if that is the issue, there is a much better algorithm.